private/pkg/manifest/manifest.go

```diff
@@ -0,0 +1,289 @@
+
+func newManifestError(lineno int, msg string) *ManifestError {
+	return &ManifestError{
+		lineno: lineno,
+		msg:    msg,
+	}
+}
+func newManifestErrorFromError(lineno int, err error) *ManifestError {
+	return &ManifestError{
+		lineno:  lineno,
+		msg:     err.Error(),
+		wrapped: err,
+	}
+}
+
+func (e *ManifestError) Error() string {
+	return fmt.Sprintf("invalid manifest: %d: %s", e.lineno, e.msg)
+}
+
+func (e *ManifestError) Unwrap() error {
+	return e.wrapped
+}
+
+// Manifest represents a list of pathToDigest and their digests.
+type Manifest struct {
+	pathToDigest  map[string]Digest
+	digestToPaths map[string][]string
+	hash          sha3.ShakeHash
+}
+
+var _ encoding.TextMarshaler = (*Manifest)(nil)
+var _ encoding.TextUnmarshaler = (*Manifest)(nil)
+
+// NewManifest creates an empty manifest.
+func NewManifest() *Manifest {
+	return &Manifest{
+		pathToDigest:  make(map[string]Digest),
+		digestToPaths: make(map[string][]string),
+		hash:          sha3.NewShake256(),
+	}
+}
+
+func splitManifest(data []byte, atEOF bool) (int, []byte, error) {
+	// Return a line without LF.
+	if i := bytes.IndexByte(data, '\n'); i >= 0 {
+		return i + 1, data[0:i], nil
+	}
+
+	// EOF occurred with a partial line.
+	if atEOF && len(data) != 0 {
+		return 0, nil, errNoFinalNewline
+	}
+
+	return 0, nil, nil
+}
+
+// NewManifestFromReader builds a manifest an encoded manifest reader.
+func NewManifestFromReader(manifest io.Reader) (*Manifest, error) {
+	m := NewManifest()
+	scanner := bufio.NewScanner(manifest)
+	scanner.Split(splitManifest)
+	lineno := 0
+	for scanner.Scan() {
+		lineno++
+		hash, path, found := strings.Cut(scanner.Text(), "  ")
+		if !found {
+			return nil, newManifestError(lineno, "invalid entry")
+		}
+		digest, err := NewDigestFromString(hash)
+		if err != nil {
+			return nil, newManifestErrorFromError(lineno, err)
+		}
+		if err := m.addDigest(path, digest); err != nil {
+			return nil, newManifestErrorFromError(lineno, err)
+		}
+	}
+	err := scanner.Err()
+	if err == errNoFinalNewline {
+		return nil, newManifestError(lineno, "partial record")
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return m, nil
+}
+
+// NewManifestFromBucket creates a manifest from a bucket.
+func NewManifestFromBucket(
+	ctx context.Context,
+	bucket storage.ReadBucket,
+) (*Manifest, error) {
+	m := NewManifest()
+	err := bucket.Walk(ctx, "", func(info storage.ObjectInfo) error {
+		path := info.Path()
+		obj, err := bucket.Get(ctx, path)
+		if err != nil {
+			return err
+		}
+		if err := m.AddContent(path, obj); err != nil {
+			return err
+		}
+		return obj.Close()
+	})
+	if err != nil {
+		return nil, err
+	}
+	return m, nil
+}
+
+func (m *Manifest) addDigest(path string, digest *Digest) error {
+	if digest.Type() != shake256Name {
+		return fmt.Errorf("unsupported hash: %s", digest.Type())
+	}
+	if n := len(digest.Bytes()); n != shake256Length {
+		return fmt.Errorf("short digest: %d of %d bytes", n, shake256Length)
```

Comment:
Nit related to [this previous comment](https://github.com/bufbuild/buf/pull/1499#discussion_r999582015), it's erroring with `"short"` but the condition is `!=`, I'd use `different|unexpected`.